frontend/appflowy_tauri/src/appflowy_app/components/document/NumberedListBlock/NumberedListBlock.hooks.ts

```diff
@@ -0,0 +1,26 @@
+import { useAppSelector } from '$app/stores/store';
+import { BlockType, NestedBlock } from '$app/interfaces/document';
+
+export function useNumberedListBlock(node: NestedBlock<BlockType.NumberedListBlock>) {
+  // Find the last index of the previous blocks
+  const prevNumberedIndex = useAppSelector((state) => {
+    const nodes = state['document'].nodes;
+    const children = state['document'].children;
+    // The parent must be existed
+    const parent = nodes[node.parent!];
+    const siblings = children[parent.children];
+    const index = siblings.indexOf(node.id);
+    if (index === 0) return 0;
+    const prevNodeIds = siblings.slice(0, index);
+    // The index is distance from last block to the last non-numbered-list block
+    const lastIndex = prevNodeIds.reverse().findIndex((id) => {
+      return nodes[id].type !== BlockType.NumberedListBlock;
+    });
+    if (lastIndex === -1) return prevNodeIds.length;
+    return lastIndex;
+  });
+
+  return {
+    index: prevNumberedIndex + 1,
+  };
+}
```

Comment:
```ts
    const { nodes, children } = useAppSelector((state) => state['document']);
    const parent = nodes[node.parent!];
    const siblings = children[parent.children];
    const index = siblings.indexOf(node.id);
    if (index === 0) return 1;
    const prevNodeIds = siblings.slice(0, index);
    // The index is distance from last block to the last non-numbered-list block
    const lastIndex = prevNodeIds.reverse().findIndex((id) => {
      return nodes[id].type !== BlockType.NumberedListBlock;
    });
    const prevNumberedIndex = lastIndex === -1 ?  prevNodeIds.length : lastIndex;
    return {
      index: prevNumberedIndex + 1,
    };
```

It seems the state only used in the first two line. Is it necessary to wrap them in a function?

Reply:
Your code is equivalent to subscribing to all state updates, not only when the index changes